fix: asdf 0.16 breaking changes

richtong opened the pull request:
This changes the commands from .bash to executables and also changes the setup so that it uses the new asdf cmd direnv envrc syntax


## Overview


The new asdf 0.16 moves to a Go implementation and also changes the way that it calls, specifically to access direnv, you need to utter `asdf cmd direnv envrc` with the new cmd function. The fix is that the ./config/ installer has to add an extra cmd.

Also the direnv commands should no longer end in .bash and should become executable as other folks have documented in 

Fixes https://github.com/asdf-community/asdf-direnv/issues/194

Fortunately asdf just git clones this repo so no installation changes, just the files here

The old asdf bash versions disabled set -e and -o pipefail, but if you have these loaded then various scripts will fail. There is a one line change to lib-asdf (I tried to avoid it but couldn't think of a way to do it where it does a -n check for a variable that may not exist. I changed it to -v).

Also if you are using venv, the old `asdf direnv local python 3.12.7` command breaks. I fixed two problems here, first the `maybe_run` is broken because of the -e and -o pipefail problems, so I turned those off when doing maybe_run.

Now that `local` has disappeared I changed the direnv command to `set` although the local still works

Note that this assumes the legacy bash scripts in asdf are still there.

Closes #194 

## Implementation notes

How does it accomplish it, in broad strokes? i.e. How does it change the codebase?

The ./lib commands have .bash removed and are made executable

The install script adds the cmd prefix 

## Interesting/controversial decisions

Include anything that you thought twice about, debated, chose arbitrarily, etc. 
What could have been done differently, but wasn't? And why?

There will continue to be the problem of asdf removing all their scripts which this sources, so cached the current asdf bash library as of the 0.16. We could do this as a submodule and pin it instead, but it seems strange that a plugin would cache the parent. It would be better to just track changes or get them
to make the bash library a standalone library that both share.

## Test coverage

There are two problems with the tests, first is that asdf 0.16 introduces a change in syntax so all older scripts will fail, specifically it is now `asdf cmd direnv` and not `asdf direnv` I updated the readme.

Second is that the lib-asdf do not pass shellcheck as they are from a different project. I left them in there, but we may need to exclude that directory from the GitHub action as hacking at this old code will further diverse it.

## Loose ends

As noted with the deprecation of the bash library, we will have to maintain it or have to rewrite asdf-direnv

reneploetz commented:
Thanks for this!

I think we do not necessarily need to copy over the whole asdf codebase. The relevant functions are in utils.bash and versions.bash of asdf - which we could strip down even further.
This would also be consistent with other plugins that copy at least the utils.bash, e.g. https://github.com/asdf-vm/asdf-nodejs/blob/master/lib/utils.sh

Yes, the "direnv install" command depends on asdf install_tool_version and reshim_command additionally to the two filts above but I think we can just use the actual asdf commands as they are fast enough now?

richtong commented:
Agreed that is what I did. I just copied in the bash scripts in that are needed. Utils.sh needs some other files. I didn't investigate if all the bash files are needed so I just copied the whole lib sub tree. I don't know direnv internals enough to know.

Also I forgot to copy into the PR from the discussion. But this change breaks the test scripts which use the 0.15 and lower commands. I didn't change these as this felt beyond the scope.

But the test scripts need a way to detect it is asdf 0.16 or higher and use different command. They have to add cmd to everything. The easy way to do this is something like

asdf $CMD direnv

Where CMD is null if it's 0.16 and below.

Finally the README is now broken so I updated all references to note if you are using asdf 0.16 or above.

It might be nice to version the direnv plugin with a version as well so you bc a say direnv plugin v.xx requires 0.16.

I also had to hack at the asdf direnv local python 3.12.7 command. Forgot to mention that. Since local and global are now set.

It now is asdf cmd direnv set python 3.12.7

So the driven script changed. The local is still supported in the scopes

Finally the asdf scripts do not pass shell check. I didn't want to hack at these and didn't feel like I should modify the GitHub action.

Final final thing is that because it looks like asdf turned off set -e and -o pipe fail the maybe_run longer works so I changed on reference looking for as non existent ASDF_DIR to do a -v and not a -n.

And there is another part in maybe_run which does. A command || false but this fails if set -o pipefail is set.

If should really be a caught error

That is

if ! Command; then
Log effor
FI

But I just turned off pipefail for that one. Wasn't sure how you wanted to handle it. Because in general command || true doesn't work with pipefail on as I recall

Sent via Superhuman iOS ( ***@***.*** )

On Tue, Feb 11, 2025 at 00:19, René Plötz < ***@***.*** > wrote:

> 
> 
> 
> 
> 
> 
> 
> Thanks for this!
> 
> 
> 
> I think we do not necessarily need to copy over the whole asdf codebase.
> The relevant functions are in utils.bash and versions.bash of asdf - which
> we could strip down even further.
> This would also be consistent with other plugins that copy at least the
> utils.bash, e.g. https://github.com/asdf-vm/asdf-nodejs/blob/master/lib/utils.sh
> 
> 
> 
> 
> Yes, the "direnv install" command depends on asdf install_tool_version and
> reshim_command additionally to the two filts above but I think we can just
> use the actual asdf commands as they are fast enough now?
> 
> 
> 
> —
> Reply to this email directly, view it on GitHub (
> https://github.com/asdf-community/asdf-direnv/pull/195#issuecomment-2648728616
> ) , or unsubscribe (
> https://github.com/notifications/unsubscribe-auth/AANTCR6UYGYWWAZBAELDUED2PDNSDAVCNFSM6AAAAABWZS57WKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNBYG4ZDQNRRGY
> ).
> You are receiving this because you authored the thread. Message ID: <asdf-community/asdf-direnv/pull/195/c2648728616
> @ github. com>
> 
> 
>

richtong commented:
OK some additional fixes, "use asdf" is also broken, so I fixed in the code base so it properly calls "asdf cmd direnv" rather than "asdf direnv"

Since this is copied to ~/.config/direnv/lib/use_asdf.sh, everyone with an old installation also need to fix that file as

richtong commented:
Final point, if you get to this and it is not merged yet, then

"asdf install direnv https://github.com/richtong"

Has the fixes here

MXfive commented:
> Final point, if you get to this and it is not merged yet, then
> 
> "asdf install direnv [@richtong](https://github.com/richtong?rgh-link-date=2025-02-15T15%3A01%3A37Z)"
> 
> Has the fixes here

Is this suppsoed to work while this is still not merged @richtong?

Getting this:

```
> asdf install direnv @richtong
∗ Downloading and installing direnv...

Cleanup: Something went wrong!

61 /Users/martyn/.asdf/plugins/direnv/bin/install: curl --fail --silent --location --create-dirs --output "$source_path" "$download_url"
```

gotgenes commented:
> Is this suppsoed to work while this is still not merged @richtong?
> 
> Getting this:
> 
> ```
> > asdf install direnv @richtong
> ∗ Downloading and installing direnv...
> 
> Cleanup: Something went wrong!
> 
> 61 /Users/martyn/.asdf/plugins/direnv/bin/install: curl --fail --silent --location --create-dirs --output "$source_path" "$download_url"
> ```

The correct invocation is

```sh
asdf plugin add direnv https://github.com/richtong/asdf-direnv
```

faraazkhan commented:
@reneploetz this is a much needed, can we please get this merged?

jfly commented:
@faraazkhan, I don't know if @reneploetz has the power to merge this PR.

I do have the power to merge PRs, but I don't feel comfortable vendoring all this code from an old version of asdf, and I'm not really maintaining this project anymore. See https://github.com/asdf-community/asdf-direnv/issues/196. If anyone here is willing to take up maintenance, please reach out over there.

My advice would be for someone to start a conversation with upstream asdf about this. Most of this project is awkwardly hooking into asdf's internal library. If we move this into core asdf (as an alternative to shims), I hypothesize it would be a lot simpler for everyone.

reneploetz commented:
I do not have any commit permissions. Or any other permissions within this project.

In my initial comments I stated what I would feel would have to be done. Personally, I would not pull in all of the code of asdf because it just decreases the maintainability of the overall codebase in this project without much gain.
Instead, as mentioned, I would just copy over the needed functionality. Alternatively, since the go version is much improved in terms of speed, it might also be possible to implement some - if not all - functionality by calling asdf commands directly.
One of the main reasons to use this plugin - cached versions with fast reshim - is likely not much of an issue anymore since the latest asdf version seems to have been much improved here. That leaves the environment variables functionality as we as the printing of the currently used versions as main advantage. Unless somebody else has another usecase?!

On the other hand, I do not feel that I should review code without being a designated reviewer - my opinion is as good as anyone else's and I do not want to add my own opinion on something a maintainer has to decide.

As far as integrating into asdf core goes, the upstream maintainer appears to take the approach that relying or using internal data structures of asdf should not be done - which is a reasonable approach obviously. See also: https://github.com/asdf-vm/asdf/pull/1937#issuecomment-2665857345
I also think that asking to integrate direnv into a go project might be a tall order - which is why I think the most reasonable approach is in fact calling asdf commands and use their output.

As this is a maintainer decision I do not really have a say here - so this are just my two cents.

dagadbm commented:
I think I will try this: https://mise.jdx.dev/

and if it works i might just stop using asdf all together and be done with
it.



On Tue, Mar 11, 2025 at 4:19 PM René Plötz ***@***.***> wrote:

> I do not have any commit permissions. Or any other permissions within this
> project.
>
> In my initial comments I stated what I would feel would have to be done.
> Personally, I would not pull in all of the code of asdf because it just
> decreases the maintainability of the overall codebase in this project
> without much gain.
> Instead, as mentioned, I would just copy over the needed functionality.
> Alternatively, since the go version is much improved in terms of speed, it
> might also be possible to implement some - if not all - functionality by
> calling asdf commands directly.
> One of the main reasons to use this plugin - cached versions with fast
> reshim - is likely not much of an issue anymore since the latest asdf
> version seems to have been much improved here. That leaves the environment
> variables functionality as we as the printing of the currently used
> versions as main advantage. Unless somebody else has another usecase?!
>
> On the other hand, I do not feel that I should review code without being a
> designated reviewer - my opinion is as good as anyone else's and I do not
> want to add my own opinion on something a maintainer has to decide.
>
> As far as integrating into asdf core goes, the upstream maintainer appears
> to take the approach that relying or using internal data structures of asdf
> should not be done - which is a reasonable approach obviously. See also: asdf-vm/asdf#1937
> (comment)
> <https://github.com/asdf-vm/asdf/pull/1937#issuecomment-2665857345>
> I also think that asking to integrate direnv into a go project might be a
> tall order - which is why I think the most reasonable approach is in fact
> calling asdf commands and use their output.
>
> As this is a maintainer decision I do not really have a say here - so this
> are just my two cents.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/asdf-community/asdf-direnv/pull/195#issuecomment-2714953907>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD7WGRYW3H77EPSPZROY6JD2T4EKLAVCNFSM6AAAAABWZS57WKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMJUHE2TGOJQG4>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
> [image: reneploetz]*reneploetz* left a comment
> (asdf-community/asdf-direnv#195)
> <https://github.com/asdf-community/asdf-direnv/pull/195#issuecomment-2714953907>
>
> I do not have any commit permissions. Or any other permissions within this
> project.
>
> In my initial comments I stated what I would feel would have to be done.
> Personally, I would not pull in all of the code of asdf because it just
> decreases the maintainability of the overall codebase in this project
> without much gain.
> Instead, as mentioned, I would just copy over the needed functionality.
> Alternatively, since the go version is much improved in terms of speed, it
> might also be possible to implement some - if not all - functionality by
> calling asdf commands directly.
> One of the main reasons to use this plugin - cached versions with fast
> reshim - is likely not much of an issue anymore since the latest asdf
> version seems to have been much improved here. That leaves the environment
> variables functionality as we as the printing of the currently used
> versions as main advantage. Unless somebody else has another usecase?!
>
> On the other hand, I do not feel that I should review code without being a
> designated reviewer - my opinion is as good as anyone else's and I do not
> want to add my own opinion on something a maintainer has to decide.
>
> As far as integrating into asdf core goes, the upstream maintainer appears
> to take the approach that relying or using internal data structures of asdf
> should not be done - which is a reasonable approach obviously. See also: asdf-vm/asdf#1937
> (comment)
> <https://github.com/asdf-vm/asdf/pull/1937#issuecomment-2665857345>
> I also think that asking to integrate direnv into a go project might be a
> tall order - which is why I think the most reasonable approach is in fact
> calling asdf commands and use their output.
>
> As this is a maintainer decision I do not really have a say here - so this
> are just my two cents.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/asdf-community/asdf-direnv/pull/195#issuecomment-2714953907>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD7WGRYW3H77EPSPZROY6JD2T4EKLAVCNFSM6AAAAABWZS57WKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMJUHE2TGOJQG4>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>
